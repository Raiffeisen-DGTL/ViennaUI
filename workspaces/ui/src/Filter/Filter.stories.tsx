import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import Filter, { FilterProps } from './Filter';

export default {
    title: 'Development/Filter',
    component: Filter,
} as Meta;

export const Overview: Story<FilterProps> = (args) => {
    return (
        <Filter {...args}>
            <Filter.Item>Пол</Filter.Item>
            <Filter.Item>Рост</Filter.Item>
            <Filter.Item>Вес</Filter.Item>
        </Filter>
    );
};
