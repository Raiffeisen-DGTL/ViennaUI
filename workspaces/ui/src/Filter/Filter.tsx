import React, { useCallback } from 'react';
import { Item, Props as ItemProps } from './Item';
import { Groups, GroupsProps } from '../Groups';

interface Props extends Omit<GroupsProps, 'size' | 'design'> {
    /** Дизайн */
    design?: 'primary' | 'ghost';
    /** Выравнивание */
    align?: 'horizontal' | 'vertical';
    /** Размер */
    size?: 'm';
}

export const Filter: React.FC<React.PropsWithChildren<Props>> & {
    Item: React.FC<React.PropsWithChildren<ItemProps>>;
} = (props: React.PropsWithChildren<Props>) => {
    const { children, size = 'm', design = 'primary', align = 'horizontal', ...attrs } = props;

    const constructChildren = useCallback(
        () => React.Children.toArray(children).map((child: any) => React.cloneElement(child, { size, design })),
        [children, size, design]
    );

    return (
        <Groups design={align} {...attrs}>
            {constructChildren()}
        </Groups>
    );
};

Filter.displayName = 'Filter';
Filter.Item = Item;
export default Filter;
