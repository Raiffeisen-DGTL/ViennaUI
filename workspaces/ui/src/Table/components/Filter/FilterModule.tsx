import { Module } from '../../types';

export type FilterState = Record<string, any>;

export const FilterModule: Module = {
    name: 'filter',
    initState: ({ settings }): FilterState => {
        return settings.filter;
    },
};
