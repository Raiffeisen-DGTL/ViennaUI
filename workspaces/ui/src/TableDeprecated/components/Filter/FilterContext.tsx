import React, { createContext, useContext } from 'react';

export const FilterContext = createContext({});

export const FilterProvider = ({ id, children }) => {
    return <FilterContext.Provider value={id}>{children}</FilterContext.Provider>;
};

export const useFilterContext = () => {
    const id = useContext(FilterContext);
    if (id === undefined) {
        throw new Error('useFilterContext must be used within a TableProvider');
    }

    return id;
};
