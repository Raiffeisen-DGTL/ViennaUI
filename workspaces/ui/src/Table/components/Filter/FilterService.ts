import { ServiceFactory } from '../TableService/ServiceFactory';
import { FilterModule, FilterState } from './FilterModule';

export interface FilterService {
    getFilter: () => FilterState | undefined;
    getFilterColumn: (id) => any;
    setFilter: (id, value) => void;
}

export const filterService: ServiceFactory<FilterService> = function (getState, updateState) {
    const getFilter = () => {
        return getState().filter;
    };
    const getFilterColumn = (id) => {
        return getState().filter?.[id];
    };

    const setFilter = (id, value) => {
        updateState(FilterModule.name, (prev) => {
            const filter = prev.filter ?? {};
            filter[id] = value;

            return { ...prev, filter };
        });
    };

    return { getFilter, getFilterColumn, setFilter };
};
