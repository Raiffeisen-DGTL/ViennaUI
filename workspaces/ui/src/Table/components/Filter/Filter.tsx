import React, { useRef, useContext, useMemo } from 'react';
import { ThemeContext } from 'styled-components';
import { Groups, Popover } from 'vienna.ui';
import { ThemeProvider, getPresets } from 'vienna.ui-primitives';
import { Filter as FilterIcon } from 'vienna.icons';
import { SortInFilter, SortIcon } from '../Sort';
import { useTableService, useTableConfig } from '../Context';
import { ColumnTitle } from '../ColumnTitle';
import { FilterProvider } from './FilterContext';
import { TableLayers } from '../../TableLayers';
import { ColumnProps } from '../Column';
import { TableProps } from '../../Table';

export interface TableFilterProps {
    column: ColumnProps;
    size?: TableProps['size'];
}

const presets = getPresets('table.filterPopover', {
    wrapper: null,
    base: null,
    size: 'size',
});

const buildTheme = (props) => {
    const theme = {
        popover: {
            message: {
                custom: {
                    ...presets.base(props),
                    ...presets.size(props),
                },
            },
            wrapper: {
                custom: {
                    ...presets.wrapper(props),
                },
            },
        },
    };

    return theme;
};

const toggleParentTh = (ref, isOpen) => {
    return () => {
        if (ref.current && ref.current.closest('th')) {
            const zIndex = isOpen ? `${TableLayers.ActiveFilter}` : '';
            ref.current.closest('th').style.zIndex = zIndex;
        }
    };
};

let popoverRef;
const popoverRefGetter: any = (ref) => {
    popoverRef = ref;
};
const close = () => {
    if (popoverRef) {
        popoverRef.close();
    }
};

export const Filter: React.FC<TableFilterProps> = (props) => {
    const { column } = props;
    const { id, title, sortable } = column;
    const { setFilter, getFilterColumn, getSortColumn } = useTableService();
    const {
        columns: { filters },
    } = useTableConfig();
    const filter = filters?.get(id);
    const titleRef = useRef<HTMLDivElement>(null);

    const setFilterCb = React.useCallback(
        (value) => {
            setFilter(id, value);
        },
        [setFilter]
    );

    const isActiveFilter = getFilterColumn(id) !== undefined && getFilterColumn(id) !== null;
    const isActiveSort = getSortColumn()?.field === id;
    const isActive = isActiveFilter || isActiveSort;
    const icon = isActiveFilter || !isActiveSort ? <FilterIcon size='s' /> : <SortIcon field={id} />;

    const content = React.useMemo(() => {
        const filterContent = typeof filter === 'function' ? filter(setFilterCb) : filter;
        return (
            <Groups design='vertical'>
                <ColumnTitle title={title} active={isActive} icon={icon} onClick={close} forceIconVisibility />
                <div style={{ flexGrow: 1 }}>{filterContent}</div>
                {sortable && <SortInFilter field={id} />}
            </Groups>
        );
    }, [id, title, isActive, icon, filter, setFilterCb]);

    const themedContext = useContext(ThemeContext);
    const theme = useMemo(() => {
        return buildTheme({ ...props, theme: themedContext });
    }, [props, themedContext]);

    return (
        <div ref={titleRef}>
            <FilterProvider id={id}>
                <ThemeProvider theme={theme}>
                    <Popover
                        onOpen={toggleParentTh(titleRef, true)}
                        onClose={toggleParentTh(titleRef, false)}
                        anchor='bottom'
                        content={content}
                        width={248}
                        noClose
                        ref={popoverRefGetter}>
                        <ColumnTitle title={title} active={isActive} icon={icon} />
                    </Popover>
                </ThemeProvider>
            </FilterProvider>
        </div>
    );
};
