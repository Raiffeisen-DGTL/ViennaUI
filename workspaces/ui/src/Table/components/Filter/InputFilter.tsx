import React, { useCallback } from 'react';
import { Input } from '../../../index';
import { useTableService } from '../Context';
import { useFilterContext } from './FilterContext';

export const InputFilter = (props) => {
    const id = useFilterContext();
    const { getFilterColumn, setFilter } = useTableService();
    const value = getFilterColumn(id) ?? '';

    const onChange = useCallback(
        (_, { value }) => {
            setFilter(id, value || undefined);
        },
        [setFilter]
    );

    return <Input {...props} value={value} size='s' onChange={onChange} />;
};
