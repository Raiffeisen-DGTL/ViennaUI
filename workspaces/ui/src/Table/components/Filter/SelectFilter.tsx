import React, { useCallback } from 'react';
import { Select } from '../../../index';
import { useTableService } from '../Context';
import { useFilterContext } from './FilterContext';

export const SelectFilter = (props) => {
    const id = useFilterContext();
    const { getFilterColumn, setFilter } = useTableService();
    const value = getFilterColumn(id);

    const onChange = useCallback(
        (_, { value }) => {
            setFilter(id, value);
        },
        [setFilter]
    );

    return <Select {...props} value={value} fitOptions={false} size='s' onSelect={onChange} />;
};
