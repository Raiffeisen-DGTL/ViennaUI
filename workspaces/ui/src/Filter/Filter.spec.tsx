import React from 'react';
import Filter from './Filter';

test('Alarm w/ primary', async () => {
    const snap = snapshot.render(
        <Filter>
            <Filter.Item>Пол</Filter.Item>
            <Filter.Item maxWidth={100}>Рост очень высокий</Filter.Item>
            <Filter.Item>Вес</Filter.Item>
        </Filter>
    );
    expect(snap).toMatchSnapshot();
});

test('Alarm w/ ghost', async () => {
    const snap = snapshot.render(
        <Filter design='ghost'>
            <Filter.Item>Пол</Filter.Item>
            <Filter.Item maxWidth={100}>Рост очень высокий</Filter.Item>
            <Filter.Item>Вес</Filter.Item>
        </Filter>
    );
    expect(snap).toMatchSnapshot();
});
