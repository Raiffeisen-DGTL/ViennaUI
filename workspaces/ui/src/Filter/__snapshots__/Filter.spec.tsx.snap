// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alarm w/ ghost 1`] = `
<div
  class="Groupsstyles__Box-sc-1dief2o-0 fGINAc"
>
  <div
    class="Groupsstyles__Item-sc-1dief2o-1 htkhjN"
  >
    <div
      class="Itemstyles__Box-sc-1dxtwiv-0 juVrZJ"
    >
      <div
        class="Itemstyles__Label-sc-1dxtwiv-1 lpUFm"
      >
        Пол
      </div>
      <div
        class="Itemstyles__Icon-sc-1dxtwiv-2 jfGbwv"
        tabindex="1"
      >
        <svg
          fill="currentColor"
          height="16"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <path
            d="M12 13.414l-7.293 7.293-1.414-1.414L10.586 12 3.293 4.707l1.414-1.414L12 10.586l7.293-7.293 1.414 1.414L13.414 12l7.293 7.293-1.414 1.414L12 13.414z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    class="Groupsstyles__Item-sc-1dief2o-1 htkhjN"
  >
    <div
      class="Itemstyles__Box-sc-1dxtwiv-0 juVrZJ"
    >
      <div
        class="Itemstyles__Label-sc-1dxtwiv-1 haXSsx"
      >
        Рост очень высокий
      </div>
      <div
        class="Itemstyles__Icon-sc-1dxtwiv-2 jfGbwv"
        tabindex="1"
      >
        <svg
          fill="currentColor"
          height="16"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <path
            d="M12 13.414l-7.293 7.293-1.414-1.414L10.586 12 3.293 4.707l1.414-1.414L12 10.586l7.293-7.293 1.414 1.414L13.414 12l7.293 7.293-1.414 1.414L12 13.414z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    class="Groupsstyles__Item-sc-1dief2o-1 htkhjN"
  >
    <div
      class="Itemstyles__Box-sc-1dxtwiv-0 juVrZJ"
    >
      <div
        class="Itemstyles__Label-sc-1dxtwiv-1 lpUFm"
      >
        Вес
      </div>
      <div
        class="Itemstyles__Icon-sc-1dxtwiv-2 jfGbwv"
        tabindex="1"
      >
        <svg
          fill="currentColor"
          height="16"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <path
            d="M12 13.414l-7.293 7.293-1.414-1.414L10.586 12 3.293 4.707l1.414-1.414L12 10.586l7.293-7.293 1.414 1.414L13.414 12l7.293 7.293-1.414 1.414L12 13.414z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`Alarm w/ primary 1`] = `
<div
  class="Groupsstyles__Box-sc-1dief2o-0 fGINAc"
>
  <div
    class="Groupsstyles__Item-sc-1dief2o-1 htkhjN"
  >
    <div
      class="Itemstyles__Box-sc-1dxtwiv-0 kYIhvW"
    >
      <div
        class="Itemstyles__Label-sc-1dxtwiv-1 lpUFm"
      >
        Пол
      </div>
      <div
        class="Itemstyles__Icon-sc-1dxtwiv-2 jfGbwv"
        tabindex="1"
      >
        <svg
          fill="currentColor"
          height="16"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <path
            d="M12 13.414l-7.293 7.293-1.414-1.414L10.586 12 3.293 4.707l1.414-1.414L12 10.586l7.293-7.293 1.414 1.414L13.414 12l7.293 7.293-1.414 1.414L12 13.414z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    class="Groupsstyles__Item-sc-1dief2o-1 htkhjN"
  >
    <div
      class="Itemstyles__Box-sc-1dxtwiv-0 kYIhvW"
    >
      <div
        class="Itemstyles__Label-sc-1dxtwiv-1 haXSsx"
      >
        Рост очень высокий
      </div>
      <div
        class="Itemstyles__Icon-sc-1dxtwiv-2 jfGbwv"
        tabindex="1"
      >
        <svg
          fill="currentColor"
          height="16"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <path
            d="M12 13.414l-7.293 7.293-1.414-1.414L10.586 12 3.293 4.707l1.414-1.414L12 10.586l7.293-7.293 1.414 1.414L13.414 12l7.293 7.293-1.414 1.414L12 13.414z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    class="Groupsstyles__Item-sc-1dief2o-1 htkhjN"
  >
    <div
      class="Itemstyles__Box-sc-1dxtwiv-0 kYIhvW"
    >
      <div
        class="Itemstyles__Label-sc-1dxtwiv-1 lpUFm"
      >
        Вес
      </div>
      <div
        class="Itemstyles__Icon-sc-1dxtwiv-2 jfGbwv"
        tabindex="1"
      >
        <svg
          fill="currentColor"
          height="16"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <path
            d="M12 13.414l-7.293 7.293-1.414-1.414L10.586 12 3.293 4.707l1.414-1.414L12 10.586l7.293-7.293 1.414 1.414L13.414 12l7.293 7.293-1.414 1.414L12 13.414z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
